internal/pkg/levee/levee.go

```diff
@@ -79,16 +102,114 @@ func run(pass *analysis.Pass) (interface{}, error) {
 	return nil, nil
 }
 
+func doPointerAnalysis(pass *analysis.Pass, analysisConf *config.Config, c *ssa.Call) (analyzedSuccesfully bool) {
+	// Pointer analysis requires at least one main package containing a main function.
+	// If we fail to provide such a package, the analysis will fail and produce error output in stderr.
+	pkg := c.Parent().Pkg
+	main := pkg.Func("main")
+	if main == nil {
+		return false
+	}
+
+	// This configuration is used to tell the pointer package how we want the analysis to be performed.
+	// For our purposes, we just need to give it a main package and a set of values for which we want
+	// the points-to set to be computed.
+	pointerConf := &pointer.Config{
+		Mains: []*ssa.Package{pkg},
+	}
+
+	// In general, we can't use the call's arguments as values directly. For example, when calling a variadic
+	// function, the variadic argument will be a slice, but we are not interested in the slice's points-to set,
+	// rather we are interested in the points-to sets of the values in the slice.
+	vs := callValues(c)
+	if len(vs) == 0 {
+		return false
+	}
+
+	for _, v := range vs {
+		// If we ask the pointer package to analyze a value it can't analyze (i.e., not a pointer or
+		// pointer-like type), the analysis will fail and produce error output in stderr.
+		if !pointer.CanPoint(v.Type()) {
+			continue
+		}
+		// Let the configuration know that we are interested in computing this value's points-to set.
+		// The points-to set is the set of values that this pointer value might point to.
+		pointerConf.AddQuery(v)
+	}
+
+	// Perform the actual analysis, in accordance with our configuration.
+	// The pointer package's documentation says that "an error can occur only due to an internal bug".
+	// However, failing to provide a main package, or querying for a non-queryable value, will both lead
+	// to errors. We prevent these cases with explicit checks above.
+	result, err := pointer.Analyze(pointerConf)
+	if err != nil {
+		return false
+	}
+
+	// A call was analyzed successfully if performing pointer analysis on it led to a report.
+	analyzedSuccesfully = false
+	for _, ptr := range result.Queries {
+		// Obtain the points-to set for this pointer. Labels are used to get at the actual ssa.Values
+		// that are pointed to.
+		labels := ptr.PointsTo().Labels()
+		if len(labels) == 0 {
+			continue
+		}
+		for _, lab := range labels {
+			v := lab.Value()
+			switch t := v.(type) {
+			case *ssa.MakeInterface:
+				v = t.X
+			}
+			if !analysisConf.IsSource(utils.Dereference(v.Type())) {
+				continue
+			}
+			analyzedSuccesfully = true
+			reportAtSource(pass, v.(ssa.Node), (ssa.Node)(c))
```

Comment:
Perhaps, instead of reporting, we should emit a Fact, and later check for the presence of sanitizers?

Reply:
We could, but I think the "checking for sanitizers" part would require a lot of work, and in fact would likely require us to implement things that we were hoping pointer analysis would do for us, and that if we do implement, would obviate the need to use pointer analysis. For example:
```
func CallSinkWrapperWrapper(s core.Source) {
    SinkWrapperWrapper(s)
}
func SinkWrapperWrapper(i interface{}) {
    safe := Sanitize(i)
    SinkWrapper(safe)
}
func SinkWrapper(i interface{}) {
    core.Sink(i)
}
func Sanitize(i interface{}) interface{} {
    return i
}
```
`pointer` analysis would tell us that the `i` argument to `core.Sink` could point to the source `s` defined in `CallSinkWrapperWrapper`. However, that value has been properly sanitized. But `pointer` analysis doesn't tell us anything about the path the value may have traveled, so we have no easy way to know that. Having the call graph would tell us where to look, and we might be able to reconstruct that path (there could easily be many paths, though), but then we would need to fully analyze the bodies of every function on the path from source to sink in order to piece together the path the value followed, and see that it was sanitized before reaching the sink.

Reply:
If there are no sanitizers on the CallGraph then we can safely report a finding, correct?
However, if there is one - we need to analyze bodies of the functions on the call-graph to ensure that the sink was dominated by a sanitizer, correct?